Lwt 5.0.0 - promises and concurrent I/O; Lwt_ppx 2.0.0

aantron opened the pull request:
From the [changelog](https://github.com/ocsigen/lwt/releases/tag/5.0.0):

> Breaking
>
> *See ocsigen/lwt#584 for an extended summary and discussion of this release as a whole, or individual issues for each change specifically.*
>
> - The callback passed to [`Lwt.async`](http://ocsigen.org/lwt/dev/api/Lwt#VALasync) must now evaluate to `unit Lwt.t`, rather than `_ Lwt.t` (ocsigen/lwt#603, requested @cfcs).
> - [`Lwt.choose`](http://ocsigen.org/lwt/dev/api/Lwt#VALchoose), [`Lwt.nchoose`](http://ocsigen.org/lwt/dev/api/Lwt#VALnchoose), [`Lwt.nchoose_split`](http://ocsigen.org/lwt/dev/api/Lwt#VALnchoose_split), [`Lwt.pick`](http://ocsigen.org/lwt/dev/api/Lwt#VALpick), and [`Lwt.npick`](http://ocsigen.org/lwt/dev/api/Lwt#VALnpick) now raise `Invalid_argument` if given the empty list (ocsigen/lwt#557, Tim Reinke).
> - Catch nested calls to [`Lwt_main.run`](http://ocsigen.org/lwt/dev/api/Lwt_main#VALrun) (ocsigen/lwt#607, ocsigen/lwt#609, prompted François-René Rideau).
> - Use the new [`Lwt_unix.IO_vectors`](http://ocsigen.org/lwt/dev/api/Lwt_unix.IO_vectors) in [`Lwt_unix.recv_msg`](http://ocsigen.org/lwt/dev/api/Lwt_unix#VALrecv_msg) and [`Lwt_unix.send_msg`](http://ocsigen.org/lwt/dev/api/Lwt_unix#VALsend_msg) (ocsigen/lwt#594, prompted Marcello Seri).
> - Make [`Lwt_unix.Async_switch`](http://ocsigen.org/lwt/dev/api/Lwt_unix#TYPEasync_method) a synonym for `Lwt_unix.Async_detach` (ocsigen/lwt#572).
> - Remove the redundant `configure.ml` (ocsigen/lwt#700).
> - PPX: remove support for general `[%lwt ...]` expressions (ocsigen/lwt#527).
> - PPX: remove support for `Lwt_log` and the `-log` option (ocsigen/lwt#520).
> - PPX: remove the `-no-debug` option (ocsigen/lwt#528).

camelus commented:
Commit: 0afae23af7f25512b3b981a2ef775ff08ae14976

A pull request by opam-seasoned @aantron.

##### :sunny: All lint checks passed <small>0afae23af7f25512b3b981a2ef775ff08ae14976</small>

* These packages passed lint tests: `lwt.5.0.0`, `lwt_ppx.2.0.0`


---


##### :sunny: Installability check (+0)

aantron commented:
Looking at the revdeps now.

aantron commented:
I added some constraints in https://github.com/ocaml/opam-repository/pull/15545. I'll rebase this PR after that is merged.

For packages where only the tests were broken, I disabled the tests. Wherever a new release is (weakly) needed, I ccd maintainers.

<br>

---

The following assume `Lwt.async : (unit -> 'a Lwt.t) -> unit`. Lwt 5.0.0 has `Lwt.async : (unit -> unit Lwt.t) -> unit` (ocsigen/lwt#603).

- Many releases of [`tcpip`](https://ci.ocaml.org/log/saved/docker-run-f13d6d97e21e8dac2f4eaf8f33c2f3fb/80c105416d6b3740309ce78ffa28a1824f602d20), including the latest. cc @hannesm
- The latest release of [`links`](https://ci.ocaml.org/log/saved/docker-run-524f4d77f7781465bc9d2cfd6cba93c8/c25f9cdafd8927db28a5613789a7a4f1161b5384). cc @dhil @jstolarek
- Tests of all releases of [`arp-mirage`](https://ci.ocaml.org/log/saved/docker-run-2ed6876b9e908fa0883dac056ccb1f84/5861fedd3d7dff8e2499551e2c5353dfecc3d0ef).

<br>

---

The following rely on the old signature of `Lwt_send_msg`, which changed in ocsigen/lwt#594.

- [`obus` 1.2.0](https://ci.ocaml.org/log/saved/docker-run-c178391dae0018fba87af9607d29b425/8fd0768dfa30485803e1852338fc5104b65d75ae). cc @Freyr666
- Tests of recent releases of [`tuntap`](https://ci.ocaml.org/log/saved/docker-run-39d821b9a843f5a94dd7c70840f6afdd/865f118ea92c3622e438cd698975ec3ba92b4386).

<br>

---

The following make a nested call to `Lwt_main.run`, which was never intended to work. Since ocsigen/lwt#603, Lwt raises `Failure` on such a nested call.

- Tests of recent releases of [`mirage-logs`](https://ci.ocaml.org/log/saved/docker-run-d6fe747387ca7c514582e50d58f937bc/c843447088a76df54145dfea39018366790a8f15).
- Tests of [`noise`](https://ci.ocaml.org/log/saved/docker-run-ea7a96ce52fbbedca73dec3957926915/c96dae9e75b7937e4896394064f23de2c82e2354).

<br>

---

The following depend on `lwt.unix` and link to it through `ocamlfind`, but don't pass the `-threads` flag. This is actually an unintentional breaking change, not related to the fact that this is a major release (it could have accidentally occurred in any release). It is caused by ocsigen/lwt#609.

- Tests of the latest releases of [`lwt-binio`](https://ci.ocaml.org/log/saved/docker-run-7d50fa3e66d9025f5a80937c74235b2d/cd64fff9c034f373ac00d6ac96371396f74c2200). This package builds its tests unconditionally, however, so I added constraints instead of only disabling tests (i.e., there is no separate test process to disable). cc @hcarty
- Tests of the latest releases of [`lwt-zmq`](https://ci.ocaml.org/log/saved/docker-run-f200a43cc11324345b5637f758f8d884/8633b3686181a47e7e07f9a95e2a3202750af245). This package also has tests built with the main build, so I added constraints. cc @hcarty
- Recent releases of [`named-pipe`](https://ci.ocaml.org/log/saved/docker-run-f3bff200ab574482ae263e4600ca3707/34ea7402a91df70c85a51098ebbb734e18bf8fce). cc @hannesm
- [`nbd` 3.0.0](https://ci.ocaml.org/log/saved/docker-run-d6ab95aca24b972c5053ef0d9ec9cfda/ca2c3b247f79510fde6d8826dea766d106f7617d). cc @gaborigloi
- [`tlstunnel` 0.2.0](https://ci.ocaml.org/log/saved/docker-run-18ea29290340a5d9116242c755e6559f/fe8ae89677613bef037d423dc61460c26d96ed9e). cc @hannesm
- Tests of [`zbar`](https://ci.ocaml.org/log/saved/docker-run-01d366841b4cc4bbe7b1d567b954348b/c5688c79db95716f251b41da4445afba6e8aacd0). These appear to be part of the main build, so I added a constraint. cc @vbmithr
- Tests of older releases of [`inotify`](https://ci.ocaml.org/log/saved/docker-run-89c7e5d5f36c17660e4da7483c6cf292/5cbcaa4b7ef8038aa5acd084a68f0177d34f7719). This is not a problem for the latest release of `inotify`.
- Tests of older releases of [`irc-client`](https://ci.ocaml.org/log/saved/docker-run-19a8da1096dcd22f28c9af85d7b3c608/b19d3e894e2975e5bc00685015a84d914bff568d). Newer releases of `irc-client` don't depend on Lwt, and `irc-client-lwt` doesn't have the issue.
- Tests of [`markup` 0.7.5](https://ci.ocaml.org/log/saved/docker-run-8d4c5fab729d44f114db46cdcb7dce44/8b7822476b30280583e08325257a8a475e2176b4). Newer releases don't have the issue.
- Tests of [`mirage-clock-unix` 1.2.0](https://ci.ocaml.org/log/saved/docker-run-22ad59061e55cad03284c07668dfe5a9/77c94607fbe540199d24e41566b56f11e2b0e54a). Newer releases don't have the issue.
- Tests of [`mirage-flow-unix` 1.2.0](https://ci.ocaml.org/log/saved/docker-run-59c0d2406d108b22110216684c0dea9b/2d1c95376352c6979142694d2f8706c2502277e3). Newer releases don't have the issue.
- Tests of [`mirage-logs` 1.3.0](https://ci.ocaml.org/log/saved/docker-run-f8c8a1dbca6d9041ee4ffcf84e51fa9c/6e6c37ec12b131b3746aaf8cd0097cba7a17764a). Newer releases don't have the issue.
- Tests of recent releases of [`ocurl`](https://ci.ocaml.org/log/saved/docker-run-c8df3891fd4054dca829e5e303446b9e/4fb28af9f4376fd77a77af5167bb3fc2cf943e21).
- Tests of [`osc` 0.1.4](https://ci.ocaml.org/log/saved/docker-run-55fc8818700109d6126c675eddaad5af/51c27dd150af9f77372156ef6ab5ca1e5ebaaf80).
- Tests of recent releases of [`tls`](https://ci.ocaml.org/log/saved/docker-run-916a4fad838907ad46183a9cd7edf4fa/81332ad954083309f7858bf58b19b091a2ea378a).
- Tests of [`xen-evtchn` 1.0.7](https://ci.ocaml.org/log/saved/docker-run-28dda176df3e8261589fc2514bd20b07/d5f0ffcc119c0959cb992d7901a5ee260b95462d). These appear to be built as part of the ordinary library build, so I added a constraint. Newer releases are available that are not affected.

<br/>

---

`brotli` 1.3.0 failed for a reason unrelated to Lwt. I'll submit a separate PR for that.

dhil commented:
> The latest release of links. cc @dhil @jstolarek

Not sure whether this information is useful, but: A new release of Links (based off HEAD) is due in January (tentatively 31st January). The fix is already [present in master](https://github.com/links-lang/links/commit/d8323d07538aa83613770d9a3fe0d34580980e26).

hannesm commented:
@aantron thanks, I have PRs for arp-mirage (tests) and tls (lwt examples, should unblock tlstunnel as well). will merge and release this week hopefully

//cc @mirage/core anyone feels responsible for tcpip (I think there's a PR pending), nbd (no idea), mirage-logs (tests using nested Lwt_main.run), named-pipe (no idea)?

talex5 commented:
re mirage-logs, I believe this is a bug in Lwt. Here's a simpler test case that also fails with Lwt 5.0.0:

```ocaml
#require "lwt";;
#require "lwt.unix";;

let () =
  try Lwt_main.run (Lwt.fail_with "hi")
  with _ -> Lwt_main.run Lwt.return_unit
```

Produces:

```
$ utop a.ml
Exception: Failure "Nested calls to Lwt_main.run are not allowed".
Fatal error: exception Failure("Nested calls to Lwt_main.run are not allowed")
```

aantron commented:
Thanks @talex5, I will fix and re-tag.

Freyr666 commented:
Thank you for the reminder, I'll do a new `obus` release as soon as possible.

kit-ty-kate commented:
Is this error expected?
```
# File "_none_", line 1:
# Error: No implementations provided for the following modules:
#          Mutex referenced from /home/opam/.opam/4.09/lib/lwt/unix/lwt_unix.cmxa(Lwt_main)
# Command exited with code 2.
```

aantron commented:
No. It looks like opam is not respecting conflict `with-test` constraints:

```
conflicts: [
  "lwt" {with-test & >= "5.0.0"}
]
```

It should have downgraded Lwt for the tests, or not tested this package at all. The package was included in the constraints PR.

aantron commented:
All the revdeps failures except two are either not due to Lwt, or due to this apparent issue with conflicts and `with-test`. The remaining two are new constraints I should add:

- The tests of [`mirage-logs` 0.3.0](https://ci.ocaml.org/log/saved/docker-run-55b37162f358977deca52e0ca9667828/c381ff9e573d090a294bd709aa304b0ff246f20f), which also have the issue with `-threads`.
- [`notty` 0.2.1](https://ci.ocaml.org/log/saved/docker-run-1c8bec250a4a47a57d58c5f73f138a25/77f31a2a6c1675414e3e1b286239bc5e579f8ab1), which isn't compatible with `Lwt.async : (unit -> unit Lwt.t) -> unit`. `notty` has newer releases that are. Indeed, IIRC, it was notty which triggered the request to change to this signature in Lwt.

I suggest to merge this PR after we decide what to do about conflicts and `with-test`, if anything, and I will add constraints on `mirage-logs` tests and `notty` in a follow-on PR.

kit-ty-kate commented on packages/lwt_ppx/lwt_ppx.2.0.0/opam:
```diff
@@ -0,0 +1,34 @@
+dev-repo: "git+https://github.com/ocsigen/lwt.git"
+
+depends: [
+  "dune"
```
```suggestion
  "dune" {>= "1.1"}
```
According to its dune-project file, this package requires at least dune 1.1

aantron replied:
Indeed, in fact this package's `dune` file works with Dune 1.0, but it shares `dune-project` with package `lwt`.

kit-ty-kate commented:
I've asked the opam maintainers about `with-tests` in the conflicts field, we'll see what we can do

aantron commented:
I just tested this locally with an open repo that contains only `markup.0.7.5`. When I added:

```
conflicts: [
  "lwt" {>= "4.5.0"}
]
```

the result was

```
Installing lwt.4.4.0.
∗ installed lwt.4.4.0
Installing markup.0.7.5.
∗ installed markup.0.7.5
Done.
```

When I changed that to

```
conflicts: [
  "lwt" {with-test & >= "4.5.0"}
]
```

the result was

```
Installing lwt.4.5.0.
∗ installed lwt.4.5.0
Installing markup.0.7.5.
∗ installed markup.0.7.5
```

In both cases, the command was

```
opam install --dry-run markup.0.7.5 lwt --with-test
```

This seems like a bug in opam.

aantron commented:
In my view, this PR is ready for merge:

- The `with-test` issue is reported in ocaml/opam#4048 (thanks!). All the affected packages already have constraints in their `opam` files, so the constraints will either start working in a future opam release, or we can easily find them if they need to be modified again.
- The remaining 5.0.0 constraints are merged (#15570). I don't see a need for another rebase of this PR, though we could do it, for one last check. I doubt we will find any new packages.
